PX4/failsafe-overhaul.md

```diff
@@ -0,0 +1,136 @@
+- Feature Name: failsafe_overhaul
+- Start Date: 2022-07-19
+- RFC PR: [dronecode/rfc#0000](https://github.com/dronecode/rfc/pull/0000)
+
+# Summary
+[summary]: #summary
+
+The current implementation of failsafe actions in PX4 is very rigid and limited. It is impossible to combine failsafe conditions, trigger multiple actions, or set progressively more aggressive actions (currently hardcoded fore battery voltage monitoring). It is also very difficult to extend the failsafe system to respond to user specific conditions because it is tightly tied into `commander`. Generalizing failsafe configuration to a standard framework with variable conditions and actions allows more flexibility. Additionally, more conditions and actions should be added.
+
+There is a fixed set of failsafe “conditions” each with an associated action. This code is mostly handled in the commander module. This leaves limited room to implement more complex/realistic failsafe scenarios, including combining conditions, and progressively more aggressive actions.
+
+# Motivation
+[motivation]: #motivation
+
+The current failsafe system lacks features that my team needed and I believe can be useful to other people, such as:
+* Progressive failsafes. Battery monitoring has a warning, failsafe, and emergency level hardcoded. But this feature would be very useful for other situations, such as telemetry loss for 30 seconds vs 5 minutes.
+* Combining failsafe actions. The action taken when something fails can depend on a combination of conditions. If RC is lost and telemetry still exists or vice versa, it might be safe to continue the mission - or RTL, but if RC is lost and telemetry also becomes lost, the drone is not within communication and it may be best to land. This is a simple example, but a feature like this allows an operator to define precisely the best course of action for any combination of failures.
+* Multiple actions.
+* Additional actions and triggers, including sending messages over MAVLink to the companion computer or over CAN bus. 
```

Comment:
Can you be more specific here?

Reply:
@bkueng what part should I clarify/expand? This entire chunk or just line 19?